pkg/dependencies/factory.go

```diff
@@ -636,8 +642,39 @@ func (f *Factory) WithNetworking(clusterConfig *v1alpha1.Cluster) *Factory {
 	return f
 }
 
+func (f *Factory) WithCNIInstaller(spec *cluster.Spec, provider providers.Provider) *Factory {
```

Comment:
It's unfortunate that we need to pass the provider, which makes dep construction a two step process.

We only need the namespaces from it. My suggestion would be to decouple this data retrieval from the provider (it's static and known at build time). This would mean having another switch for the datacenter type, but I think it's worth it. The questions would be, where do we put it. The `providers` package would be ideal, but that would create a circular dependency if we try to import the provider specific packages from there.